compiler/rustc_passes/src/dead.rs

```diff
@@ -59,21 +59,22 @@ fn struct_all_fields_are_public(tcx: TyCtxt<'_>, id: DefId) -> bool {
     // treat PhantomData and positional ZST as public,
     // we don't want to lint types which only have them,
     // cause it's a common way to use such types to check things like well-formedness
-    tcx.adt_def(id).all_fields().all(|field| {
-        let field_type = tcx.type_of(field.did).instantiate_identity();
-        if field_type.is_phantom_data() {
-            return true;
-        }
-        let is_positional = field.name.as_str().starts_with(|c: char| c.is_ascii_digit());
-        if is_positional
-            && tcx
-                .layout_of(tcx.param_env(field.did).and(field_type))
-                .map_or(true, |layout| layout.is_zst())
-        {
-            return true;
-        }
-        field.vis.is_public()
-    })
+    tcx.has_attr(id, sym::externally_constructed)
```

Comment:
Is this a good place for this check? In function `check_item`, around line 838 you have a variable `may_construct_self` and in `visit_node` there is `unconditionally_treats_fields_as_live`

Reply:
I think it's good, cause `struct_all_fields_are_public` is also used in `create_and_seed_worklist`.

Reply:
Oh, seems checking it in `create_and_seed_worklist` is good enough. And I opened another PR which just checks the `repr(c)` or `repr(transparent)`, see #127104.